examples/apps/src/ble_bas_peripheral.rs

```diff
@@ -110,40 +97,42 @@ async fn ble_task<C: Controller>(mut runner: Runner<'_, C>) {
     }
 }
 
-/// Run the Gatt Server.
-async fn gatt_task(server: &Server<'_, '_>) {
-    loop {
-        if let Err(e) = server.run().await {
-            #[cfg(feature = "defmt")]
-            let e = defmt::Debug2Format(&e);
-            panic!("[gatt_task] error: {:?}", e);
-        }
-    }
-}
-
 /// Stream Events until the connection closes.
-async fn conn_task(server: &Server<'_, '_>, conn: &Connection<'_>) -> Result<(), Error> {
+async fn conn_task(server: &Server<'_>, conn: &Connection<'_>) -> Result<(), Error> {
     let level = server.battery_service.level;
     loop {
         match conn.next().await {
             ConnectionEvent::Disconnected { reason } => {
                 info!("[gatt] disconnected: {:?}", reason);
                 break;
             }
-            ConnectionEvent::Gatt { event, .. } => match event {
-                GattEvent::Read { value_handle } => {
-                    if value_handle == level.handle {
-                        let value = server.get(&level);
-                        info!("[gatt] Read Event to Level Characteristic: {:?}", value);
+            ConnectionEvent::Gatt { data } => {
+                // We can choose to handle event directly without an attribute table
+                // let req = data.request();
+                // ..
+                // data.reply(conn, Ok(AttRsp::Error { .. }))
+
+                // But to simplify things, process it in the GATT server that handles
+                // the protocol details
+                match data.process(server.deref()).await {
```

Comment:
I thought so too,  but somehow the dyn trait is not allowing me to
```
error[E0277]: the trait bound `Server<'_>: trouble_host::prelude::DynamicAttributeServer` is not satisfied
   --> /home/lulf/dev/trouble/examples/apps/src/ble_bas_peripheral.rs:117:36
    |
117 |                 match data.process(server).await {
    |                                    ^^^^^^ the trait `trouble_host::prelude::DynamicAttributeServer` is not implemented for `Server<'_>`
    |
    = help: the trait `trouble_host::prelude::DynamicAttributeServer` is implemented for `trouble_host::prelude::AttributeServer<'_, M, MAX>`
    = note: required for the cast from `&Server<'_>` to `&dyn trouble_host::prelude::DynamicAttributeServer`
```

I tried to modify the deref as well, but same issue:

```
            impl<'values> core::ops::Deref for #name<'values>
            {
                type Target = dyn DynamicAttributeServer + 'values;

                fn deref(&self) -> &Self::Target {
                    &self.server
                }
            }
```

Reply:
Ohh interesting......yeah I see what you mean.

I had a bit of a play around and it seems like the issue stems from the fact `server` is of type `&Server<'_>` rather than `Server<'_>`. So explicitly invoking the impl'd `Deref` trait with `data.process(&**server).await` is fine by the compiler, but it's certainly not intuitive from a user perspective.

I'm just going to throw some thoughts out there to properly explore this:

First off, what is the reason for the `DynamicAttributeServer` trait, rather than `process` and the like just taking `&AttributeServer` as an argument instead? Are you thinking of implementing other types of attribute server in future? Or is the purpose just to stop external use of the methods?

I feel like maybe the function should be called something other than `deref()` as I don't think it's immediately clear to the user that that is what they'd need to call.

Probably the most user friendly approach would be for `process` and the other similar functions to accept a reference to `Server` directly, and accessing the `AttributeServer` can be done within the function. The issue with this of course is that the `Server` type is macro-generated, making the function signature definition tricky. I think we would need a `GattServer` trait which is automatically implemented on the generated `Server`, and have the functions accept `&impl GattServer` as their argument. (I'm suggesting static dispatch as in most use cases I'd expect there would be only one server defined). I think this function signature would be very intuitive for the user though.

Another option would be to rename `deref()` to `attribute_server()` or something. Also explains what's going on, but I think for a user who doesn't need to know the nuts and bolts of what the library is doing it's less clear than the other option.